go/vt/vtctl/grpcvtctldserver/server.go

```diff
@@ -5173,11 +5173,32 @@ func (s *VtctldServer) ApplyKeyspaceRoutingRules(ctx context.Context, req *vtctl
 	span.Annotate("skip_rebuild", req.SkipRebuild)
 	span.Annotate("rebuild_cells", strings.Join(req.RebuildCells, ","))
 
-	if err := s.ts.SaveKeyspaceRoutingRules(ctx, req.KeyspaceRoutingRules); err != nil {
+	resp := &vtctldatapb.ApplyKeyspaceRoutingRulesResponse{}
+
+	// We send both the old and new rules back in the response, so that we have a backup of the rules
+	// in case we need to revert them or for audit purposes.
+	currentRules, err := s.ts.GetKeyspaceRoutingRules(ctx)
```

Comment:
This Old / New behavior is non-standard for the topo server and for Vitess. Older versions of the key are stored in the topo server if it came to that.

Can you help me understand what is different/special about KeyspaceRoutingRules that requires or warrants this Old / New behavior? I didn't see anything related to this in the issue or PR description. If it's truly warranted — then I think we should lay out why in the PR description, along with showing an example command and output.

Reply:
The reason I thought of this is not special to keyspace routing rules. Just that with the number of workflows being quite high (depending on number of tenants) the probability of wanting to update routing rules might be higher than for other migration flows.  

Actually I had originally thought of changing the `Apply()` for keyspace routing rules to only update one tenant at a time, but it didn't match what the other `Apply`s do: they modify the entire rule file. We decided we didn't want a feature creep and could implement this later if required.

In fact it may not be a bad idea for most user-facing RPCs/commands to the topo done by users to send back previous state as well since it makes it easy to audit/backup. It is a pattern I have used often in the past for "admin" commands.

>  Older versions of the key are stored in the topo server if it came to that.

Do we have a way to access a key value today from the topo or do you mean using the underlying topo's mechanisms like `etcdctl get`? In that case, depending on the key, we may need post-processing of the value
before displaying or using it for an `Apply` to revert a previous update.

I agree though, for DR purposes, it is a good option to have and I hadn't thought of it: we do have intrinsic versioning! But for regular users sending the previous version provides a lightweight way to audit/backup older rules..

Again, I can change it to not send the old rules and let the onus lie on the user doing a `Get` before an `Apply`.  Let me know what you all want.

Reply:
I'm not set against it, I just think the we should document — in the PR description and the code — why we're doing it since it's unique within Vitess today AFAIK.

I also think it would be better/nicer to add a command flag to control this: ~ `--include-previous-value` (even if it defaults to `true`). We could then use that elsewhere going forward too when/where it might be nice.

Reply:
The model I'm using for these `Apply` command is `ApplySchema` / `ApplyVSchema`, which in fact return the "new" state. I'm not sure why `ApplyRoutingRules` did not follow that pattern, and it seems like at least initially, the `ApplyRoutingRules` pattern was then followed by `ApplyShardRoutingRules` and `ApplyKeyspaceRoutingRules`. To make them all consistent, we should at least return the "new" state.
Let us not feature-creep this PR into adding a `--version` to `GetTopologyPath`, because then the next thing we'll need to add is a way of listing versions with enough metadata to be able to choose a version to inspect.
Either returning just the "new" value, or returning both behind a flag are fine by me.

Reply:
We should definitely return the *new* state. What's unusual is returning the old and new. 

I agree that we shouldn't expand the scope of the work here, I was only noting what I think would be a nicer way to achieve what I believe to be the aim here with returning the old value. Adding e.g. `--version` to `GetTopologyPath` would IMO be a much nicer and more thorough way to address what I perceived as the reasoning for it — but it should definitely be a separate piece of work (we/I could create an issue for it though).